File: core/src/com/spacefist/managers/ProjectileManager.java
package com.spacefist.managers;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import com.spacefist.GameData;
import com.spacefist.ai.projectilebehaviors.SeekingBehavior;
import com.spacefist.entities.Entity;
import com.spacefist.entities.Projectile;
import com.spacefist.entities.SpaceBlock;
import com.spacefist.entities.enemies.Enemy;
import org.jetbrains.annotations.NotNull;

/// <summary>
/// Keeps track of the projectiles in the world.
/// </summary>
public class ProjectileManager extends Manager<Projectile> {

    /// <summary>
    /// Creates a new ProjectileManager instance
    /// </summary>
    /// <param name="gameData">Common game data</param>
    public ProjectileManager(GameData gameData) {
        super(gameData);
    }

    @Override
    public void Update() {

        for (Projectile projectile : entities) {
            Rectangle resolution = gameData.getResolution();

            Rectangle rect = new Rectangle(
                (int)gameData.getCamera().x,
                (int)gameData.getCamera().y,
                resolution.getWidth(),
                resolution.getHeight()
            );

            // Mark offscreen live projectiles as dead and
            // only update onscreen projectiles.
            if (projectile.isAlive())
            {
                if (rect.contains(projectile.getRectangle()))
                {
                    projectile.update();
                }
                else
                {
                    projectile.setAlive(false);
                }
            }
        }
    }

    /// <summary>
    /// Places a laser at the specified location and fires it.
    /// </summary>
    /// <param name="x">The X value of the location</param>
    /// <param name="y">The Y value of the location</param>
    public void fireLaser(int x, int y)
    {
        fireLaser(x, y, new Vector2(0, -1), false);
    }

    /// <summary>
    /// Fires a laser from the specified location in the specified direction.
    /// </summary>
    /// <param name="x">The X value of the location</param>
    /// <param name="y">The Y value of the location</param>
    /// <param name="direction">The direction to send the projectile</param>
    /// <param name="enemyLaser">Whether this laser belongs to an enemy</param>
    public void fireLaser(int x, int y, @NotNull Vector2 direction, boolean enemyLaser)
    {
        gameData.getRoundData().setShotsFired(gameData.getRoundData().getShotsFired() + 1);

        float rotation = ((float) Math.toDegrees((float) Math.atan2(direction.y, direction.x)));

        // Place a new active laser at x, y
        Projectile projectile = new Projectile(
            gameData,
            gameData.getTextures().get("Laser"),
            new Vector2(x, y),
            direction,
            9,
            enemyLaser
        );

        projectile.setRotation((float) Math.toRadians(rotation));

        Add(projectile);

    }

    /// <summary>
    /// Fires a rocket cluster from the specified location
    /// </summary>
    /// <param name="x">The X component of the location</param>
    /// <param name="y">The Y component of the location</param>
    public void fireSampleWeapon(int x, int y)
    {
        Iterable<Enemy> visibleEnemies     = gameData.getEnemyManager().getVisibleEnemies();
        Iterable<SpaceBlock> visibleBlocks = gameData.getBlockManager().getVisibleBlocks();

        Array<Entity> onScreen = new Array<Entity>(false, 16);

        onScreen.addAll((Array<? extends Entity>) visibleEnemies);
        onScreen.addAll((Array<? extends Entity>) visibleBlocks);


        // TODO: fireSampleWeapon: ConcurrentModification Bug
        for(Entity entity : onScreen) {
            if(entity.getY() >= y) {
                onScreen.removeValue(entity, true);
            }
        }

        if (onScreen.size != 0)
        {
            // TODO: fireSampleWeapon: IndexOutOfBounds Bug
            // Mark several onscreen entities as targets
            // and send rockets to intercept them.
            for (int i = 0; i < 4; i++)
            {
                int idx = MathUtils.random(onScreen.size);

                Entity target = onScreen.get(idx);

                // Convert tinting code in ProjectileManager
                // Mark targeted entities by tinting them red
                // target.Tint = Color.Crimson;

                gameData.getRoundData().shotFired();



                Projectile projectile = new Projectile(
                    gameData,
                    gameData.getTextures().get("SampleWeapon"),
                    new Vector2(x, y),
                    new Vector2(0, -1),
                    10,
                    false
                );

                    projectile.setBehavior(new SeekingBehavior(
                        new Vector2(0, -1),
                        new Vector2(x, y),
                        target
                    ));

                Add(projectile);
            }
        }
    }

    /*******Dongcai**********/
    public void fireBluelaser(int x, int y)
    {
        gameData.getRoundData().shotFired();

        Projectile projectile = new Projectile(
            gameData,
            gameData.getTextures().get("Mine"),
            new Vector2(x, y),
            new Vector2(0, -1),
            -2,
            false
        );

        Add(projectile);
    }

    public void fireMissile(int x, int y)
    {
        gameData.getRoundData().shotFired();

        Projectile projectile = new Projectile(
            gameData,
            gameData.getTextures().get("Missile"),
            new Vector2(x, y),
            new Vector2(0, -1),
            20,
            false
        );

        Add(projectile);

        Projectile projectile1 = new Projectile(
            gameData,
            gameData.getTextures().get("Missile"),
            new Vector2(x + 50, y),
            new Vector2(0, -1),
            10,
            false
        );

        Add(projectile1);

        Projectile projectile2 = new Projectile(
            gameData,
            gameData.getTextures().get("Missile"),
            new Vector2(x - 50, y),
            new Vector2(0, -1),
            10,
            false
        );

        Add(projectile2);

    }
    /***************************************/

    /// <returns>
    /// Returns a collection of all of the live projectiles
    /// fired by the player.
    /// </returns>
    @NotNull
    public Iterable<Projectile> PlayerProjectiles()
    {
        Array<Projectile> playerProjs = new Array<Projectile>(false, 16);

        for(Projectile projectile : entities) {
            if ((!projectile.isEnemyProjectile()) && projectile.isAlive()) {
                playerProjs.add(projectile);
            }
        }

        return playerProjs;
    }

    /// <summary>
    /// Returns a collection of all of the live projectiles
    /// fired by an enemy.
    /// </summary>
    /// <returns></returns>
    @NotNull
    public Iterable<Projectile> EnemyProjectiles()
    {
        Array<Projectile> enemyProjs = new Array<Projectile>(false, 16);

        for(Projectile projectile : entities) {
            if (projectile.isEnemyProjectile() && projectile.isAlive()) {
                enemyProjs.add(projectile);
            }
        }

        return enemyProjs;
    }
}

File: core/src/com/spacefist/managers/ProjectileManager.kt
package com.spacefist.managers

import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.Array
import com.spacefist.GameData
import com.spacefist.entities.Entity
import com.spacefist.entities.Projectile

/**
 *  Keeps track of the projectiles in the world.
 *
 *  Creates a new ProjectileManager instance
 *
 *  @param name gameData Common game data
 */
class ProjectileManager

(gameData: GameData) : Manager<Projectile>(gameData) {

    override fun update() {
        entities
            .filter  { entity -> entity.isAlive }

            .forEach { projectile ->
                val resolution = gameData.resolution

                val rect = Rectangle(
                        gameData.camera.x.toInt().toFloat(),
                        gameData.camera.y.toInt().toFloat(),
                        resolution.getWidth(),
                        resolution.getHeight()
                )

                if (rect.contains(projectile.rectangle)) {
                    projectile.update()
                } else {
                    projectile.isAlive = false
                }
            }
        }


    /**
     *  Fires a laser from the specified location in the specified direction.
     * 
     *  @param x The X value of the location
     *  @param y The Y value of the location
     *  @param direction The direction to send the projectile
     *  @param enemyLaser Whether this laser belongs to an enemy
     */
    @JvmOverloads
    fun fireLaser(x: Int, y: Int, direction: Vector2 = Vector2(0f, -1f), enemyLaser: Boolean = false) {
        gameData.roundData.shotsFired = gameData.roundData.shotsFired + 1

        val rotation = Math.toDegrees(Math.atan2(direction.y.toDouble(), direction.x.toDouble()).toFloat().toDouble()).toFloat()

        // Place a new active laser at x, y
        val projectile = Projectile(
                gameData,
                gameData.textures["Laser"]!!,
                Vector2(x.toFloat(), y.toFloat()),
                direction,
                9,
                enemyLaser
        )

        projectile.rotation = Math.toRadians(rotation.toDouble()).toFloat()

        add(projectile)

    }

   

    fun fireMissile(x: Int, y: Int) {
        gameData.roundData.shotFired()

        val projectile = Projectile(
                gameData,
                gameData.textures["Missile"]!!,
                Vector2(x.toFloat(), y.toFloat()),
                Vector2(0f, -1f),
                20,
                false
        )

        add(projectile)

        val projectile1 = Projectile(
                gameData,
                gameData.textures["Missile"]!!,
                Vector2((x + 50).toFloat(), y.toFloat()),
                Vector2(0f, -1f),
                10,
                false
        )

        add(projectile1)

        val projectile2 = Projectile(
                gameData,
                gameData.textures["Missile"]!!,
                Vector2((x - 50).toFloat(), y.toFloat()),
                Vector2(0f, -1f),
                10,
                false
        )

        add(projectile2)

    }

    /**
     *  @return Returns a collection of all of the live projectiles
     *  fired by the player.
     */
    fun playerProjectiles(): Iterable<Projectile> {
        val playerProjs = Array<Projectile>(false, 16)

        entities
                .filter  { !it.isEnemyProjectile && it.isAlive }
                .forEach { playerProjs.add(it)                 }

        return playerProjs
    }

    /**
     *  Returns a collection of all of the live projectiles
     *  fired by an enemy.
     * 
     */
    fun enemyProjectiles(): Iterable<Projectile> {
        val enemyProjs = Array<Projectile>(false, 16)

        entities
                .filter  { it.isEnemyProjectile && it.isAlive }
                .forEach { enemyProjs.add(it)                 }

        return enemyProjs
    }
}